clarity/src/vm/types/serialization.rs

```diff
@@ -452,275 +544,444 @@ impl Value {
-                        TypeSignature::SequenceType(SequenceSubtype::BufferType(expected_len)) => {
-                            u32::from(&buffer_len) <= u32::from(expected_len)
-                        }
-                        _ => false,
-                    };
-                    if !passed_test {
-                        return Err(SerializationError::DeserializeExpected(x.clone()));
-                    }
-                }
-
-                let mut data = vec![0; u32::from(buffer_len) as usize];
-
-                r.read_exact(&mut data[..])?;
+        let mut stack = vec![DeserializeStackItem::TopLevel {
+            expected_type: top_expected_type.cloned(),
+        }];
 
-                Value::buff_from(data).map_err(|_| "Bad buffer".into())
-            }
-            TypePrefix::BoolTrue => {
-                check_match!(expected_type, TypeSignature::BoolType)?;
-                Ok(Bool(true))
-            }
-            TypePrefix::BoolFalse => {
-                check_match!(expected_type, TypeSignature::BoolType)?;
-                Ok(Bool(false))
+        while !stack.is_empty() {
+            if stack.len() > MAX_TYPE_DEPTH as usize {
+                return Err(CheckErrors::TypeSignatureTooDeep.into());
             }
-            TypePrefix::PrincipalStandard => {
-                check_match!(expected_type, TypeSignature::PrincipalType)?;
-                StandardPrincipalData::deserialize_read(r).map(Value::from)
-            }
-            TypePrefix::PrincipalContract => {
-                check_match!(expected_type, TypeSignature::PrincipalType)?;
-                let issuer = StandardPrincipalData::deserialize_read(r)?;
-                let name = ContractName::deserialize_read(r)?;
-                Ok(Value::from(QualifiedContractIdentifier { issuer, name }))
-            }
-            TypePrefix::ResponseOk | TypePrefix::ResponseErr => {
-                let committed = prefix == TypePrefix::ResponseOk;
-
-                let expect_contained_type = match expected_type {
-                    None => None,
-                    Some(x) => {
-                        let contained_type = match (committed, x) {
-                            (true, TypeSignature::ResponseType(types)) => Ok(&types.0),
-                            (false, TypeSignature::ResponseType(types)) => Ok(&types.1),
-                            _ => Err(SerializationError::DeserializeExpected(x.clone())),
-                        }?;
-                        Some(contained_type)
-                    }
-                };
 
-                let data = Value::inner_deserialize_read(r, expect_contained_type, depth + 1)?;
-                let value = if committed {
-                    Value::okay(data)
-                } else {
-                    Value::error(data)
+            let expected_type = stack
+                .last()
+                .expect("FATAL: stack.last() should always be some() because of loop condition")
```

Comment:
`stack.last()`'s type is `Option<&DeserializeStackItem>` -- it will always be `Some(&DeserializeStackItem)` because the loop condition is:

```rust
while !stack.is_empty() {
...
}
```

`expected_type`, however _could_ be None, in exactly the situation (and some others) you mention.